### DISCUSSION TITLE
Unable to use any hooks

### DISCUSSION THREAD

**Discussion by suman-kr** [vercel/next.js#34258] [Help] [answered]
Hi, I'm trying to use hooks inside  a functional component which is a HOC in NextJS.

I'm getting this error
```
Error: Invalid hook call. Hooks can only be called inside of the body of a function component. This could happen for one of the following reasons:
1. You might have mismatching versions of React and the renderer (such as React DOM)
2. You might be breaking the Rules of Hooks
3. You might have more than one copy of React in the same app
See https://reactjs.org/link/invalid-hook-call for tips about how to debug and fix this problem.
```

My code

```tsx
import { useRouter } from 'next/router'
import React, { useState } from 'react'

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function ProtectedRoute(Component: any) {
  const router = useRouter()
  const [loggedIn, setLoggedIn] = useState(false)
  // const checkRouteAcces = () =>
  const token = localStorage.getItem('token')

  if (token) {
    setLoggedIn(true)
  }

  return class WrappedComponent extends React.Component {
    render() {
      if (loggedIn) return <Component />
      router.push('/admin')
    }
  }
}

export { ProtectedRoute }

```

I've already tried `if (typeof window !== "undefined")` check. and also `globalThis`. Nothing works

**Comment by 2stash** [comment, marked as answer]
Hi! I don't think this has anything to do specifically with nextjs, but I have a few suggestions for your code.

First, I would put your check for a valid token in useEffect. I'm not 100% sure it has to be in a useEffect hook, but I always have problems with conditional rendering based on an startup variable check if its just in an 'if" statement.  

Second, I would move your "router.push('/admin') out of the render and put it where you put the check for a token. I'm not sure it won't work in the render, but logically you don't even want to get to the render part if you already determined you don't have a token.

Third, what's up with using hooks and functional components to return a class component? I've never seen that.

Good luck!
